src/murmur/GRPCContainer.h

```diff
@@ -0,0 +1,492 @@
+				/// makes it very hard to do so without intent.
+				///
+				template<typename Idx>
+				typename std::unique_ptr<Idx, decltype(lock_holder())>
+				makeUniqueIdx(Idx& index) {
+					return std::unique_ptr<Idx, lock_holder>(
+							std::addressof(index),
+							lock_holder(this->m_Mtx)
+						);
+				}
+
+				///
+				/// helper function to convert a pair of iterators into a
+				/// list that can be safely iterated over 
+				/// without holding the lock on the container.
+				///
+				/// Boost Range is no longer used as it wasn't properly filtering.
+				///
+				/// \return `boost::container::slist<locked_type>` a singly-linked list
+				/// designed for forward iteration only.
+				template<typename It>
+				auto getLockedRange(const std::pair<It, It>& rng) -> boost::container::slist<locked_type>;
+
+			public:
+
+				using iterator = typename container_t::iterator;
+
+				///
+				/// uses a user-provided functor that returns a iterator pair
+				/// from the container to make a 'safe' range that can be iterated
+				/// over without either holding the mutex for the container or without
+				/// having check for deleted RPCCalls.
+				///
+				/// \param func a functor that takes in a const lref to this container,
+				/// and returns an iterator pair of the range it wants to go over.
+				///
+				/// \return `boost::container::slist<locked_type>` with the locked values. This
+				/// list is singly-linked, for forward iteration only.
+				///
+				template<typename Functor>
+				auto getLocked(Functor&& func) -> boost::container::slist<locked_type>;
+
+				///
+				/// gets a 'safe' range for an index on this container which
+				/// can be iterated over without any invalidation of the iterators and
+				/// all RPCCall pointers are non-nullptr std::shared_ptr
+				///
+				/// \param Idx a reference to the 'tag' of the index you wish to get
+				///
+				/// \return `boost::container::slist<locked_type>` a singly-linked
+				/// list with copies of the values you wanted.
+				///
+				template<typename Idx>
+				auto getLockedIndex(const Idx&) -> boost::container::slist<locked_type>;
+
+				/// \brief Only available if the container has an index on a serverId.
+				///
+				/// This will get the container mutex, then return to you a smart
+				/// pointer that you can use as if this was a std::map or std::multimap
+				/// depending on configuration.
+				///
+				/// \return a std::unique_ptr to that has an interface as if it
+				/// was a standard container.
+				///
+				template<typename Q = server>
+				auto getServerPtr() -> decltype(makeUniqueIdx(mi::get<Q>(container))) {
+
+					return makeUniqueIdx(mi::get<Q>(container));
+				}
+
+				/// \brief Only available if container has index on a serverId.
+				///
+				/// \return const lreference that can be used like a standard container
+				///
+				template<typename Q = server>
+				BOOST_ATTRIBUTE_NODISCARD auto getServerIndex() const -> decltype(mi::get<Q>(container)) {
```

Comment:
`BOOST_ATTRIBUTE_NODISCARD` is only available in more recent versions of Boost.
You should add a check after the includes whether this macro is defined and if it's not, define it to be empty

Reply:
This has been fixed.